Resources/Computing/PoolComputingElement.py

```diff
@@ -124,17 +124,18 @@ def submitJob(self, executableFile, proxy=None, **kwargs):
     if not res['OK']:
       self.log.error("Could not dump cfg to pilot.cfg", res['Message'])
 
+    # Here we define task kwargs: adding complex objects like thread.Lock can trigger errors in the task
+    taskKwargs = {'InnerCESubmissionType': self.innerCESubmissionType}
     if self.innerCESubmissionType == 'Sudo':
       for nUser in range(MAX_NUMBER_OF_SUDO_UNIX_USERS):
         if nUser not in self.userNumberPerTask.values():
           break
-      kwargs['NUser'] = nUser
-      kwargs['PayloadUser'] = os.environ['USER'] + 'p%s' % str(nUser).zfill(2)
-    kwargs['InnerCESubmissionType'] = self.innerCESubmissionType
+      taskKwargs['NUser'] = nUser
+      taskKwargs['PayloadUser'] = os.environ['USER'] + 'p%s' % str(nUser).zfill(2)
```

Comment:
Unless DIRAC forces it somewhere, it's possible for `USER` to not be set:
```suggestion
      taskKwargs['PayloadUser'] = os.environ.get('USER', 'NOUSER') + 'p%s' % str(nUser).zfill(2)
```

Reply:
I don't have much knowledge about the `SudoComputingElement`. Do you know why the username should be followed by `p + nUser`?
Is `NOUSER` something specific to this CE?

From what I understand, `PayloadUser` is passed as a CE parameter to the `SudoComputingElement`:
https://github.com/DIRACGrid/DIRAC/blob/integration/Resources/Computing/PoolComputingElement.py#L48

And then it is needed to run sudo commands: 
https://github.com/DIRACGrid/DIRAC/blob/integration/Resources/Computing/SudoComputingElement.py#L136

I assume that setting it to `NOUSER` would fail at this point (unless there is something specific with this word with these CEs).
Would it be okay to set it to `None`?

We would have: 

```
user = os.environ.get('USER')
if user:
  taskKwargs['PayloadUser'] = user + 'p%s' % str(nUser).zfill(2)
```

And then the `SudoComputingElement` would just try to use `BaseUsername`:
https://github.com/DIRACGrid/DIRAC/blob/integration/Resources/Computing/SudoComputingElement.py#L62

(Well, and after that if `BaseUsername` is not valid, it tries to get the `USER` environment variable again, may be I should add a line to stop the execution at this point)

Reply:
> I don't have much knowledge about the SudoComputingElement. Do you know why the username should be followed by p + nUser?

This is something that @Andrew-McNab-UK should be able to answer you

> Is NOUSER something specific to this CE?

No, just a "random" string here.

> I assume that setting it to NOUSER would fail at this point (unless there is something specific with this word with these CEs).
Would it be okay to set it to `None`?

I also think it would fail and that `None` would be better

Reply:
It's to use a different UID for each payload. Rather like pool accounts on worker nodes. It's tied very closely to the Pilot VM definition in the Pilot user_data file, which precreates all these accounts inside the VM. (Namespaces or full blown containers would be a better way to do this nowadays.)

Reply:
Alright thanks, is this better now?